fairscale/nn/misc/flatten_params_wrapper.py

```diff
@@ -220,41 +241,50 @@ def flat_param(self) -> nn.Parameter:
         assert len(self.flat_params) == 1, "Incorrect access to flat_param"
         return self.flat_params[0]
 
-    def _init_flatten_params(self, p_set: Set[Tuple[nn.Module, str]]) -> List[nn.Parameter]:
+    def _init_flatten_params(
+        self, p_set: Set[Tuple[nn.Module, str]]
+    ) -> Tuple[
+        List[nn.Parameter], List[Tuple[str, nn.Module, str]], List[Tuple[str, str, nn.Module, str, nn.Module, str]]
+    ]:
         """ Build metadata for need-to-be-flatten parameters and returns a list
             contains the need-to-be-flatten parameters.
 
-            This extends self._param_infos and self._shared_param_infos lists.
+            This also returns param_infos and shared_param_infos, which
+            will be attached to the flat parameter object.
 
         Args:
             p_set (set):
                 A set of (module, param_name) for a set of params that needed
                 to be flattened. There could be shared params in this set.
         """
         param_infos = []
-        shared_param_memo: Dict[nn.Parameter, Tuple[nn.Module, str]] = {}
+        shared_param_memo: Dict[nn.Parameter, Tuple[str, nn.Module, str]] = {}
         shared_param_infos = []
         params = []
         for module_name, m in self.named_modules():
             for n, p in m.named_parameters(recurse=False):
                 if p is not None and (m, n) in p_set:
                     if p in shared_param_memo:
-                        shared_m, shared_n = shared_param_memo[p]
-                        shared_param_infos.append((m, n, shared_m, shared_n))
+                        mname, shared_m, shared_n = shared_param_memo[p]
+                        shared_param_infos.append((module_name, mname, m, n, shared_m, shared_n))
                     else:
-                        shared_param_memo[p] = (m, n)
+                        shared_param_memo[p] = (module_name, m, n)
                         param_infos.append((module_name, m, n))
                         params.append(p)
         del shared_param_memo
 
-        assert len(set(p.dtype for p in params)) == 1, "expects all parameters in module to have same dtype"
+        assert len(set(p.dtype for p in params)) == 1, "expects all parameters to have same dtype"
+        assert len(set(p.requires_grad for p in params)) == 1, "expects all parameters to have same requires_grad"
```

Comment:
are you planning to skip flattening parameters if requires_grad are not the same in params?

Reply:
no, I am going to assume in the future user will just group their params in that use case. I.e. group the params with requires_grad==True together and requires_grad==False together. Then, this assert won't trigger since those will be different FlatParameters.